Add support for Pololu 3pi+ 2040 Robot and Zumo 2040 Robot

pdg137 opened the pull request:
This is to support the Pololu [3pi+ 2040 Robot](https://www.pololu.com/product/5001) and [Zumo 2040 Robot](https://www.pololu.com/category/308/zumo-2040-robot):

![image](https://github.com/micropython/micropython/assets/466760/ffbbd1bb-0f68-41a0-8e41-d3310ed85462)

![image](https://github.com/micropython/micropython/assets/466760/ee7cf71d-5058-4ef8-9e4e-b670480d26a2)

In addition to our board-specific configuration, this PR includes this change:

* For rp2 port, adds board configuration option to not have "default" I2C, SPI, and UART ports.  In our case we have multiple options for these and don't want to allow users to initialize a port without explicitly specifying pin numbers.

pdg137 commented:
FYI the new 3pi is now available [here](https://www.pololu.com/product/5001).  We are using [our own fork of micropython](https://github.com/pololu/micropython/tree/3pi) to build firmware.

Before this PR can move forward, I think we need to wait for an official pico-sdk release that supports our board, but that might still be months away; is this the normal process? We can handle maintaining our own fork indefinitely but it doesn't seem great for the community for it to take so long to add new boards.

dpgeorge commented:
> I think we need to wait for an official pico-sdk release that supports our board, but that might still be months away; is this the normal process?

It is possible to define your own board without relying on `pico-sdk`.  See for example `NULLBITS_BIT_C_PRO`, you need to set `PICO_BOARD` in the `mpconfigboard.cmake` file.

Otherwise... it would be best to wait for a release of `pico-sdk`, but I don't know how long that will be.

pdg137 commented:
Okay, I copied over that include file and rebased it retaining just the changes necessary for this board, so I think it could be merged now.  Please let me know if there's anything else you'd like to change.

pdg137 commented:
I'm not sure if anyone has looked at this yet; I'm hoping this is easy to merge and would appreciate any feedback.

dpgeorge commented on ports/rp2/machine_i2c.c:
```diff
@@ -34,6 +34,20 @@
 
 #define DEFAULT_I2C_FREQ (400000)
 
+#ifndef PICO_DEFAULT_I2C
```
If/when the config files for this robot are merged to `pico-sdk`, will they require this option to be set?  It could be that `pico-sdk` doesn't build without these defaults defined.

Because of that, I think it would be better to have a specific config option here that forces scl/sda to be provided by the user.  Eg `MICROPY_HW_I2C_NO_DEFAULT_PINS`.

pdg137 replied:
It's already merged.  I think that definition is just used in one place in pico-sdk, to define some other things:

https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_i2c/include/hardware/i2c.h#L71-L77

But sure, we can add the extra macro and maybe it will prevent problems where future board file authors just forget to define their defaults.

dpgeorge commented on shared/tinyusb/tusb_config.h:
```diff
@@ -90,7 +90,7 @@
 #define USBD_MAX_POWER_MA (250)
 
-#define USBD_DESC_STR_MAX (20)
+#define USBD_DESC_STR_MAX (40)
```
In 783ddfc264c8c36253a4279ef501f02f8cac1200 I made this configurable, so you can just increase it for your board in its `mpconfigboard.h` file.

dpgeorge commented:
Please split this PR into two commits:
- first one to force no default pins on I2C/SPI/UART
- second one to add the new board configuration files

Commit:
rp2: Add new NO_DEFAULT_PINS config options for i2c, spi, and uart.

Some boards have multiple options for these pins, and they don't want to
allow users to initialize a port without explicitly specifying pin numbers.

Signed-off-by: Paul Grayson <paul@pololu.com>

pdg137 commented:
I think it's arranged like that now. Please let me know if it looks okay!

pdg137 commented:
I rebased this to clear the conflict from https://github.com/micropython/micropython/pull/11255. Please let me know if you have any more feedback on this.

Commit:
rp2/boards: Add support for Pololu 3pi+ and Zumo robots.

Signed-off-by: Paul Grayson <paul@pololu.com>

pdg137 commented:
I updated this PR to include the (unreleased) Zumo 2040 Robot and added a picture to https://github.com/micropython/micropython-media/pull/48.

pdg137 commented:
The Zumo 2040 is now released.

We are maintaining our own MicroPython builds over at https://github.com/pololu/micropython-build but it would be great to also get this merged into the official repository.  If there's anything else I need to do for that to happen, please let me know.

dpgeorge commented:
Sorry this took a while to get back to.

Thanks for splitting up the commits.  Now merged.

pdg137 commented:
Thank you! 🎉
